### ISSUE TITLE
Default validator skips a validation batch each round

### ISSUE THREAD

**Issue by albertoveneri** [pytorch/torchtitan#2636]
### Bug description

During the default validation loop, the stop condition is checked only after a new validation batch is loaded. As a result, the validation set is reduced by one batch in each validation round.

A simple fix is to move the conditional `break` to the end of the loop, as shown in the following patch:

```bash
diff --git a/torchtitan/components/validate.py b/torchtitan/components/validate.py
--- a/torchtitan/components/validate.py
+++ b/torchtitan/components/validate.py
@@ -237,9 +237,6 @@ class Validator(BaseValidator):
         num_steps = 0
 
         for input_dict, labels in self.validation_dataloader:
-            # pyrefly: ignore [missing-attribute, unsupported-operation]
-            if self.config.steps != -1 and num_steps >= self.config.steps:
-                break
 
             self.metrics_processor.ntokens_since_last_log += labels.numel()
             for k, v in input_dict.items():
@@ -308,6 +305,10 @@ class Validator(BaseValidator):
             accumulated_losses.append(loss_sum.detach() / global_valid_tokens)
             num_steps += 1
 
+            # pyrefly: ignore [missing-attribute, unsupported-operation]
+            if self.config.steps != -1 and num_steps >= self.config.steps:
+                break
+
         # Compute average loss
         loss = torch.sum(torch.stack(accumulated_losses))
         loss /= num_steps
```

### Versions

The bug can be reproduced on the latest version of torchtitan by launching a debug run with a command similar to the following:

```bash
NGPU=1 MODULE=llama3 CONFIG=llama3_debugmodel run_train.sh --dataloader.dataset-path=tests/assets/c4_test --training.steps=100 --training.seq-len=16384 --validator.enable --validator.dataloader.dataset-path=tests/assets/c4_test --validator.steps=1 --validator.freq=1 --validator.dataloader.dataset=c4_test
```

This is an extreme case in which each validation round uses a single, large validation step. In this run, the warning message `WARNING - Dataset c4_test is being re-looped` appears only after 14 validation steps, even though fewer than 1,017 sequences have been processed out of the 2,000 sequences in [c4_test](https://github.com/pytorch/torchtitan/tree/main/tests/assets/c4_test).

To count the actual sequences processed, I applied the following patch:

```bash
diff --git a/torchtitan/components/validate.py b/torchtitan/components/validate.py
--- a/torchtitan/components/validate.py
+++ b/torchtitan/components/validate.py
@@ -136,6 +136,8 @@ class Validator(BaseValidator):
         self.pp_has_first_stage = pp_has_first_stage
         self.pp_has_last_stage = pp_has_last_stage
 
+        self.n_seq = 0  # for logging purposes only, counts number of sequences (EOS tokens) processed in validation
+
         if config.steps == -1:
             logger.warning(
                 "Setting validation steps to -1 might cause hangs because of "
@@ -304,6 +306,9 @@ class Validator(BaseValidator):
                             inputs, **extra_inputs, **extra_kwargs
                         )
                         loss_sum = self.loss_fn(predictions, labels)
+            # Count number of EOS in the batch for logging purposes
+            self.n_seq += (input_dict["input"] == self.tokenizer.eos_id).sum().item()
+            logger.info(f"Validation step {num_steps}, cumulative sequences (EOS tokens) processed: {self.n_seq}")
 
             accumulated_losses.append(loss_sum.detach() / global_valid_tokens)
             num_steps += 1
```
Moving the conditional break to the end of the loop causes the warning to appear after 28 validation steps, with 1,986 sequences read, which is expected given the large sequence length (16,384 tokens) and that each batch contains approximately 70 sequences.

**Comment by fegin** [comment]
@albertoveneri

Thanks for reporting the bug. The patch is correct to fix offset by one issue. But moving to the end of the for loop makes the readability slightly worse. Please see if https://github.com/pytorch/torchtitan/pull/2637 makes sense.
